spring-cloud-gcp-data-spanner/src/main/java/com/google/cloud/spring/data/spanner/repository/query/SqlSpannerQuery.java

```diff
@@ -232,15 +233,33 @@ private List executeReadSql(Pageable pageable, Sort sort, QueryTagValue queryTag
 
 		Statement statement = buildStatementFromQueryAndTags(queryTagValue);
 
-		return (getReturnedSimpleConvertableItemType() != null)
-				? this.spannerTemplate.query(
-						struct -> new StructAccessor(struct).getSingleValue(0), statement,
-						spannerQueryOptions)
-				: this.spannerTemplate.query(this.entityType,
-						statement,
+		if (getReturnedSimpleConvertableItemType() != null) {
+			return this.spannerTemplate.query(
+					struct -> new StructAccessor(struct).getSingleValue(0), statement,
+					spannerQueryOptions);
+		}
+		// check if returnedType is a field annotated as json
+		String jsonFieldName = getJsonFieldNameIfExists(returnedType);
+		if (jsonFieldName != null) {
+			return this.spannerTemplate.query(
+					struct -> new StructAccessor(struct).getSingleJsonValue(jsonFieldName, returnedType), statement,
```

Comment:
>If the query return type is an object like Details, is there a requirement that the entity must use Details mapped to one of the fields?

hmm, good question. I am only concerned here for cases where the the returned object is mapped to one of the fields. Like in `SELECT detail from traders where id = @id` where `detail` is already a JSON field in `traders`. If user created some new JSON type field to return in the SQL, then we have no way of knowing what object to map it to.

>Why are we using StructAccessor? Wouldn't you need something like JsonAccessor?

`StructAccessor` is used to process any query result row returned from spanner. `Struct` here does not necessarily mean a STRUCT column in spanner.

>I would imagine that all you would need to determine whether to convert from JSON is to detect that the query returns a ResultSet from Spanner with a single column that is a Type.TYPE_JSON. What am I missing here?

Right, but unlike the case of [single value returns](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/eb88c823d6d436f7359f73be0469aa98c3498df8/spring-cloud-gcp-data-spanner/src/main/java/com/google/cloud/spring/data/spanner/repository/query/SqlSpannerQuery.java#L236-L240) , I will also need to know the returnType to convert a Type.TYPE_JSON to. But you are right that I do not need `jsonFieldName` in this case. I am making changes to `SpannerPersistentEntityImpl.java` to only hold a set instead of a map for json properties. This will also resolve concerns for multiple JSON field of the same type.

Reply:
> hmm, good question. I am only concerned here for cases where the the returned object is mapped to one of the fields. Like in SELECT detail from traders where id = @id where detail is already a JSON field in traders. If user created some new JSON type field to return in the SQL, then we have no way of knowing what object to map it to.

Do you have information about the method signature, i.e. the return type of the method? That would be the type you need to map JSON too. I imagine the same problem would exist for STRUCT.

Reply:
@meltsufin You are right that the type information is available in `struct.getType().getStructFields()`. But checking the struct metadata is not a pattern used in the code base right now -- we've used it extensively for R2DBC, but Spring Data integration is much older.

@zhumin8 If the type information from Spanner is inventoried and stored in `StructAccessor` in the same way that [column names currently are](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/b29c131b675ac479658ebdd9a56fb581be405a1d/spring-cloud-gcp-data-spanner/src/main/java/com/google/cloud/spring/data/spanner/core/convert/StructAccessor.java#L159), would that be useful? On the other hand, I feel like the current way you are doing this might be more flexible, since even if the column is *stored* as JSON, the user might choose to *read* it as a `String` entity field anyway -- for example if the data needs to immediately be sent elsewhere.

Reply:
> On the other hand, I feel like the current way you are doing this might be more flexible, since even if the column is stored as JSON, the user might choose to read it as a String entity field anyway -- for example if the data needs to immediately be sent elsewhere.

@elefeint Wouldn't the user express their intent to read as `String` by specifying the return type of the query method as `String`? We can use that to decide whether to parse JSON or not.

Reply:
That's what this PR does, no?